src/domain/task-system/src/entities/task_event.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use chrono::{DateTime, Utc};
+use enum_variants::*;
+
+use super::*;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+/// All events that model life-cycle of a task
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TaskEvent {
+    /// New task entered the queue
+    Created(TaskCreated),
+    /// Task execution had started
+    Running(TaskRunning),
+    /// Cancellation of task was requested (this is not immediate and task may
+    /// still finish with a different outcome than cancelled)
+    Cancelled(TaskCancelled),
+    /// Task has reached a final outcome
+    Finished(TaskFinished),
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TaskCreated {
+    pub event_time: DateTime<Utc>,
+    pub task_id: TaskID,
```

Comment:
These 2 fields repeat in each event.
I understand that Rust has no inheritance for structs, but maybe we can have a different design:

```
struct TaskEvent {
    pub event_time: DateTime<Utc>,
    pub task_id: TaskID,
    pub body: TaskEventBody,
}

pub enum TaskEventBody {
    /// New task entered the queue
    Created(TaskCreated),
    /// Task execution had started
    Running(TaskRunning),
    /// Cancellation of task was requested (this is not immediate and task may
    /// still finish with a different outcome than cancelled)
    Cancelled(TaskCancelled),
    /// Task has reached a final outcome
    Finished(TaskFinished),
}

pub struct TaskCreated {
   pub logical_plan: LogicalPlan,
}

pub struct TaskRunning {}

pub struct TaskCancelled {}

pub struct TaskFinished {
   pub outcome: TaskOutcome,
}


```

Reply:
Yeah, this is how I started - with `task_id` and `event_time` in a separate type ... but then moved to the current design.

The reasoning is that this enum represents all events of a "Task Scheduling" domain, and there might be events that are in this domain but not directly linked to a task. 

For example in a distributed system we might want to track availability of task executors and might have `ExecutorOnline` `ExecutorOffline` events that only have `executor_id`. When executor commits to a task such event might have both `executor_id` AND `task_id`.

This series of blog posts (while a bit annoying in its style) explains this approach: https://sara.event-thinking.io/2023/04/kill-aggregate-chapter-1-I-am-here-to-kill-the-aggregate.html

Reply:
Here's a rust framework that specifically builds on the "Kill the Aggregate" idea of this blog post series https://github.com/disintegrate-es/disintegrate (one of the libraries we can evaluate later).

The idea of "events are just facts" and what we call "aggregates" are just projections reconstructed from different subsets of domain event streams really resonates with me and how I was already thinking about projections in Kamu.

Reply:
Nobody cancelled DRY principle. 
In a OOP language I would have had:
 - Event (with just event time)
    - ExecutorEvent (with executor_id)
       - subclasses
    - TaskEvent (with task_id)
       - subclasses

It's weird that we cannot express such a simple typical thing easily with Rust.